### PR TITLE
#249 Update bnd file to support CDI 1.1 to 2.x

### PR DESCRIPTION
**Pull request by arjantijms**

### CODE REVIEW

**Comment by Emily-Jiang** [comment]
I don't think we should do the changes right now. CDI 2.0  is not supported in MicroProfile as yet.

**Comment by Emily-Jiang** [comment]
Besides, the import package was worked by bnd. When MP pulls in CDI 2.0, the pom.xml will be updated with the version range on cdi dependencies, which will then generate the version range correctly in the manfest.mf. We should not change bnd.

**Comment by arjantijms** [comment]
Indeed, when you change the pom to CDI 2.0, the MANIFEST.MF will reflect that. 

The question is though, for now, do you actually want to bail out when the CDI implementation happens to be capable of supporting CDI 2, even though it in fact does support CDI 1.2 as well?

I.e. Weld 3.0 is 'reported' as CDI 2.0, but supports CDI 1.2 just perfectly fine. Only MP config will not run on it, since it explicitly says not to run on CDI 2.0.

Without this change, GlassFish 5 and Payara 5 will not be able use MP config, even though they could. The limitation seems artificial, unless I'm missing something ;)

**Comment by OndrejM** [comment]
I'm sure we have discussed support of CDI 2.0 (I think that @johnament  raised this a while ago on the mailing list). And I'm pretty sure that we decided that MP should run with future versions of any JSR as they are backwards compatible. Therefore MP should run with CDI 2.0 although it doesn't use any of its new features.

**Comment by Emily-Jiang** [comment]
Not really. In the last discussion, we decided not to auto enable cdi 2.0. I think the big enablement will come when we do MP 2.0, which should include all JavaEE8 corresponding features, JAX-RS 2.1, JSON-P 1.1, CDI 2.0. As a part of MP 2.0 release, all the current MP specs need to be updated to support the new specs. All other specs need to update their dependencies on CDI, not just this one.

**Comment by OndrejM** [comment]
@Emily-Jiang this isn't about updating dependencies for MP config - the dependency stays at CDI 1.2. But the OSGi manifest restricts any future version of CDI so it works only with a CDI implementation that provides CDI 1.2 but not with a CDI 2.0 implementation, which also provides CDI 1.2 which is a subset.

I'm strongly against such restrictions because they slow down innovation. I don't see a reason why an MP implementation can't decide to upgrade the CDI component to support CDI 2.0, while still being compatible with MP, which requires CDI 1.2 but should allow later versions.

In this [thread](https://groups.google.com/d/msg/microprofile/8Ll4PQ4Qvp8/ehty_7B0AAAJ) everybody agrees that MP should require CDI 1.2 but allow CDI 2.0. In your own words, @Emily-Jiang: " If a container has CDI 2.0 implementation, it should work with microprofile 1.0 release. "

The problem is that if a container is OSGi based and has CDI 2.0 impl, it wouldn't work with the current MP config API artifact, because the OSGi manifest doesn't allow CDI 2.0

**Comment by struberg** [comment]
Cdi will remain backward compatible.
We could easily go for [1.1,].
Means anything equals or bigger cdi 1.1

LieGrue,
Strub

**Comment by smillidge** [comment]
+1 this should be merged into the next version

**Comment by Emily-Jiang** [comment]
hang on. Please see my previous comments. The change is better to be handled in the pom.xml not here. Besides, the PR failed IP-validation.

File: api/bnd.bnd [hunk 1]
```diff
@@ -1,5 +1,10 @@
 -exportcontents: \
     org.eclipse.microprofile.*
+
+Import-Package: \
```
Comment (Emily-Jiang) [inline api/bnd.bnd]: The version range should be handled in pom.xml not here.

Comment (arjantijms) [reply api/bnd.bnd]: How would you propose handling it in pom.xml? Via the `org.apache.felix:maven-bundle-plugin` or something else?

Comment (Emily-Jiang) [reply api/bnd.bnd]: @arjantijms  yes, in maven-bundle-plugin. I did experiment a bit with pom.xml. It seems it is not as visible as specifying in the bnd.bnd. sorry, I had a false impression. What you have done is good. Actually, if cdi api was properly semantic versioned, we won't have this headache. This is why I campaign for MP to do semantic versioning properly. If you fix your ip-validation, your change can be merged in.

**Comment by Emily-Jiang** [comment]
@arjantijms can you fix the ip-validation error so that your PR can be merged?

**Comment by arjantijms** [comment]
Sorry, forgot about this one. I think I'll just have to redo the PR using my payara.fish email address. Will check for that later today. Thanks for the reminder!

**Comment by arjantijms** [comment]
PR with signed, verified and signed-off commit using @payara.fish email done at: https://github.com/eclipse/microprofile-config/pull/250

**Pull request closed without merging**